File: test/clients/ts/converge.mjs
/**
 * Scenario 7 phase runner for @opto-sync/client — cross-client convergence.
 *
 * Invoked by run_all.sh as one step of an orchestrated sequence:
 *
 *   node converge.mjs setup    # PUT the fresh fixture document
 *   node converge.mjs flush    # queue this client's payload and flush it
 *   node converge.mjs verify   # assert the final server doc + local reconcile
 *
 * Split into phases on purpose: `flush` must run once per language, in the
 * fixture's declared order, against the SAME document, so the phases cannot
 * live inside a single language's test process.
 */
import '../../../../opto-sync-clients/clients/ts/node_modules/fake-indexeddb/auto/index.mjs';
import {
  OptoSyncClient,
  SYNC_STATUS,
} from '../../../../opto-sync-clients/clients/ts/dist/index.js';

import {
  BASE_URL,
  CROSS_CLIENT as FX,
  SERVER_POLICY,
  assertDeepEqual,
  assertDeepEqualKeyed,
  getDocData,
  probeServer,
  putDoc,
  statusCounts,
  syncDoc,
} from './support.mjs';

const LANG = 'ts';
const phase = process.argv[2];
let checks = 0;

function ok(what) {
  checks += 1;
  console.log(`ok - [${LANG}/converge/${phase}] ${what}`);
}

function check(condition, what) {
  if (!condition) throw new Error(what);
  ok(what);
}

function checkEqual(actual, expected, what) {
  assertDeepEqual(actual, expected, what);
  ok(what);
}

async function setup() {
  await putDoc(FX.docId, FX.base);
  checkEqual(await getDocData(FX.docId), FX.base, `fresh document ${FX.docId} written`);
}

async function flush() {
  const payload = FX.payloads[LANG];
  if (!payload) throw new Error(`fixture has no payload for "${LANG}"`);

  // fake-indexeddb is per-process, but name the DB uniquely anyway so a rerun
  // can never observe a previous run's queue.
  const client = new OptoSyncClient({
    ...SERVER_POLICY,
    databaseName: `opto-e2e-converge-${LANG}-${Date.now()}`,
  });

  const mid = await client.queueMutation('docs', FX.docId, payload);
  check((await client.pendingMutations()).length === 1, 'payload queued as pending');

  const [queued] = await client.pendingMutations();
  const res = await syncDoc(FX.docId, JSON.parse(queued.jsonPayload));
  await client.markMutation(mid, res.ok ? SYNC_STATUS.SYNCED : SYNC_STATUS.FAILED);
  check(res.status === 200, `flushed to ${FX.docId} (HTTP ${res.status})`);
  check(res.json?.mergedWith === 'native-c-ffi', 'server merged with the native C core');

  const counts = await statusCounts(client);
  check(counts.synced === 1 && counts.pending === 0 && counts.failed === 0,
    `queue drained (${JSON.stringify(counts)})`);
  await client.db.delete();
}

async function verify() {
  const serverFinal = await getDocData(FX.docId);

  // (a) strict, order-sensitive: the server document is fully determined.
  checkEqual(serverFinal, FX.expectedFinal, 'final server document matches the predicted merge exactly');

  // Spot-check the load-bearing policy claims, so a failure names the rule.
  check(serverFinal.title === 'rust title',
    'unguarded root scalar follows arrival order (last flusher wins)');
  check(serverFinal.revision.owner === 'dart' && serverFinal.revision.updatedAt === 4000,
    'guarded object follows updatedAt, NOT flush order: rust flushed last but is stale');
  check(serverFinal.revision.priority === 2, "rust's stale revision was rejected WHOLESALE");
  // Base-only root scalar: no client payload sends a root `createdAt`, so
  // nothing can overwrite it. (This used to be attributed to FWW; `createdAt` is
  // no longer a guarded key on any tier — see SERVER_POLICY.)
  check(serverFinal.createdAt === 1000, 'base-only root createdAt untouched by every client');
  check(serverFinal.items.length === 5, 'exactly three new identities appended');
  const shared = serverFinal.items.find((i) => i.id === 'shared');
  check(shared.label === 'dart-shared' && shared.qty === 20 && shared.createdAt === 1000,
    "the shared element carries dart's write deep-merged onto the base element");
  check(serverFinal.items.find((i) => i.id === 'keep').label === 'untouched',
    'the element nobody touched is preserved verbatim');
  check(serverFinal.items.map((i) => i.id).join(',') === 'keep,shared,ts-new,dart-new,rust-new',
    'appended identities appear in flush order at the end of the array');

  // (b) this client's own local reconcile of the final state.
  const client = new OptoSyncClient({
    ...SERVER_POLICY,
    databaseName: `opto-e2e-verify-${LANG}-${Date.now()}`,
  });
  const localCopy = FX.payloads[LANG];
  const reconciled = client.reconcileIncoming('docs', FX.docId, serverFinal, localCopy);
  assertDeepEqualKeyed(reconciled, FX.expectedFinal,
    `${LANG} local reconcile of the final server state`);
  ok(`${LANG} local reconcile of the final server state agrees with every other client`);
  await client.db.delete();
}

const phases = { setup, flush, verify };

async function main() {
  if (!phases[phase]) {
    console.error(`usage: node converge.mjs <${Object.keys(phases).join('|')}>`);
    process.exit(2);
  }
  const reason = await probeServer();
  if (reason) {
    console.error(`[${LANG}] SKIP converge/${phase} — server unavailable: ${reason}`);
    process.exit(0);
  }
  await phases[phase]();
  console.log(`# [${LANG}] converge/${phase}: ${checks} checks passed against ${BASE_URL}`);
}

main().catch((err) => {
  console.error(`not ok - [${LANG}/converge/${phase}] ${err.message}`);
  process.exit(1);
});

File: test/clients/dart/bin/converge.dart
/// Scenario 7 phase runner for opto_sync_client — cross-client convergence.
///
/// Invoked by run_all.sh as one step of an orchestrated sequence:
///
///   dart run bin/converge.dart setup    # PUT the fresh fixture document
///   dart run bin/converge.dart flush    # queue this client's payload and flush it
///   dart run bin/converge.dart verify   # assert the final server doc + local reconcile
///
/// Split into phases on purpose: `flush` must run once per language, in the
/// fixture's declared order, against the SAME document, so the phases cannot
/// live inside a single language's test process.
library;

import 'dart:convert';
import 'dart:io';

import 'package:opto_sync_client/opto_sync_client.dart';
import 'package:opto_sync_client_e2e/support.dart';

const String _lang = 'dart';
late String _phase;
int _checks = 0;

void _ok(String what) {
  _checks += 1;
  stdout.writeln('ok - [$_lang/converge/$_phase] $what');
}

void _check(bool condition, String what) {
  if (!condition) throw ComparisonFailure(what);
  _ok(what);
}

void _checkEqual(dynamic actual, dynamic expected, String what) {
  expectDeepEqual(actual, expected, what);
  _ok(what);
}

Map<String, dynamic> get _payload {
  final payloads = crossClientFixture['payloads'] as Map<String, dynamic>;
  final p = payloads[_lang];
  if (p == null) throw StateError('fixture has no payload for "$_lang"');
  return jsonDecode(jsonEncode(p)) as Map<String, dynamic>;
}

String get _docId => crossClientFixture['docId'] as String;

Future<void> _setup() async {
  await putDoc(_docId, crossClientFixture['base']!);
  _checkEqual(await getDocData(_docId), crossClientFixture['base'],
      'fresh document $_docId written');
}

Future<void> _flush() async {
  final client = newClient(newSyncer());
  try {
    await client.queueMutation('docs', _docId, _payload);
    final pending = await pendingMutations(client);
    _check(pending.length == 1, 'payload queued as pending');

    final queued = pending.single;
    final res = await syncDoc(
        _docId, jsonDecode(queued.jsonPayload) as Map<String, dynamic>);
    await markMutation(
        client, queued.id, res.ok ? SyncStatus.synced : SyncStatus.failed);
    _check(res.status == 200, 'flushed to $_docId (HTTP ${res.status})');
    _check((res.json as Map)['mergedWith'] == 'native-c-ffi',
        'server merged with the native C core');

    final counts = await statusCounts(client);
    _check(counts.synced == 1 && counts.pending == 0 && counts.failed == 0,
        'queue drained ($counts)');
  } finally {
    await client.db.close();
  }
}

Future<void> _verify() async {
  final serverFinal = await getDocData(_docId);
  final expected = crossClientFixture['expectedFinal'] as Map<String, dynamic>;

  // (a) strict, order-sensitive: the server document is fully determined.
  _checkEqual(serverFinal, expected,
      'final server document matches the predicted merge exactly');

  // Spot-check the load-bearing policy claims, so a failure names the rule.
  final revision = serverFinal['revision'] as Map<String, dynamic>;
  final items = (serverFinal['items'] as List).cast<Map<String, dynamic>>();

  _check(serverFinal['title'] == 'rust title',
      'unguarded root scalar follows arrival order (last flusher wins)');
  _check(revision['owner'] == 'dart' && revision['updatedAt'] == 4000,
      'guarded object follows updatedAt, NOT flush order: rust flushed last but is stale');
  _check(revision['priority'] == 2, "rust's stale revision was rejected WHOLESALE");
  // Base-only root scalar: no client payload sends a root `createdAt`, so
  // nothing can overwrite it. (`createdAt` is no longer a guarded key on any
  // tier — FWW is a node-level veto and is opt-in.)
  _check(serverFinal['createdAt'] == 1000,
      'base-only root createdAt untouched by every client');
  _check(items.length == 5, 'exactly three new identities appended');

  final shared = items.firstWhere((i) => i['id'] == 'shared');
  _check(
      shared['label'] == 'dart-shared' &&
          shared['qty'] == 20 &&
          shared['createdAt'] == 1000,
      "the shared element carries dart's write deep-merged onto the base element");
  _check(items.firstWhere((i) => i['id'] == 'keep')['label'] == 'untouched',
      'the element nobody touched is preserved verbatim');
  _check(items.map((i) => i['id']).join(',') == 'keep,shared,ts-new,dart-new,rust-new',
      'appended identities appear in flush order at the end of the array');

  // (b) this client's own local reconcile of the final state.
  final client = newClient(newSyncer());
  try {
    final reconciled =
        await client.reconcileIncoming('docs', _docId, serverFinal, _payload);
    expectDeepEqualKeyed(reconciled, expected,
        '$_lang local reconcile of the final server state');
    _ok('$_lang local reconcile of the final server state agrees with every '
        'other client');
  } finally {
    await client.db.close();
  }
}

Future<void> main(List<String> args) async {
  final phases = <String, Future<void> Function()>{
    'setup': _setup,
    'flush': _flush,
    'verify': _verify,
  };

  _phase = args.isEmpty ? '' : args.first;
  final run = phases[_phase];
  if (run == null) {
    stderr.writeln(
        'usage: dart run bin/converge.dart <${phases.keys.join('|')}>');
    exit(2);
  }

  final reason = await probeServer();
  if (reason != null) {
    stderr.writeln('[$_lang] SKIP converge/$_phase — server unavailable: $reason');
    exit(0);
  }

  try {
    await run();
  } catch (e) {
    stderr.writeln('not ok - [$_lang/converge/$_phase] $e');
    exit(1);
  }
  stdout.writeln(
      '# [$_lang] converge/$_phase: $_checks checks passed against $baseUrl');
}
